doozer/doozerlib/backend/konflux_client.py

```diff
@@ -0,0 +1,507 @@
+        if self.dry_run:
+            self._logger.warning(f"[DRY RUN] Would have deleted {api_version}/{kind} {namespace}/{name}")
+            return
+        self._logger.info(f"Deleting {api_version}/{kind} {namespace}/{name}")
+        await exectools.to_thread(api.delete, name=name, namespace=namespace)
+        self._logger.info(f"Deleted {api_version}/{kind} {namespace}/{name}")
+
+    async def _create_or_patch(self, manifest: dict):
+        """ Create or patch a resource.
+
+        :param dyn_client: The dynamic client.
+        :param manifest: The manifest.
+        :return: The resource.
+        """
+        api_version, kind, name, namespace = self._extract_manifest_metadata(manifest)
+        resource = await self._get(api_version, kind, name, namespace, strict=False)
+        if not resource:
+            return await self._create(manifest)
+        return await self._patch(manifest)
+
+    async def _create_or_replace(self, manifest: dict):
+        """ Create or replace a resource.
+
+        :param dyn_client: The dynamic client.
+        :param manifest: The manifest.
+        :return: The resource.
+        """
+        api_version, kind, name, namespace = self._extract_manifest_metadata(manifest)
+        resource = await self._get(api_version, kind, name, namespace, strict=False)
+        if not resource:
+            return await self._create(manifest)
+        if not manifest.get("metadata", {}).get("resourceVersion"):
+            # resourceVersion is required for replace
+            manifest.setdefault("metadata", {})["resourceVersion"] = resource.metadata.resourceVersion
+        while True:
+            try:
+                return await self._replace(manifest)
+            except exceptions.ConflictError:
+                # If the resource has changed since we fetched it, retry
+                resource = await self._get(api_version, kind, name, namespace, strict=False)
+                manifest["metadata"]["resourceVersion"] = resource.metadata.resourceVersion
+                continue
+
+    @staticmethod
+    def _new_application(name: str, display_name: str) -> dict:
+        obj = {
+            "apiVersion": "appstudio.redhat.com/v1alpha1",
+            "kind": "Application",
+            "metadata": {
+                "name": name
+            },
+            "spec": {
+                "displayName": display_name,
+                "appModelRepository": {"url": ""},
+                "gitOpsRepository": {"url": ""},
+            }
+        }
+        return obj
+
+    async def ensure_application(self, name: str, display_name: str) -> resource.ResourceInstance:
+        application = self._new_application(name, display_name)
+        return await self._create_or_patch(application)
+
+    @staticmethod
+    def _new_component(name: str, application: str, component_name: str,
+                       image_repo: Optional[str], source_url: Optional[str], revision: Optional[str]) -> dict:
+        obj = {
```

Comment:
Consider storing this definition as a jinja template, instead of hardcoding it. This could make it easier to find and maintain